### PR TITLE
chore(Apps/CodeStyle): rewrite Ci-codestyle in python

### PR DESCRIPTION
**Pull request by Grimdhex**
## Changes Proposed:

This PR proposes changes to:
-  [ ] Core (units, players, creatures, game systems).
-  [ ] Scripts (bosses, spell scripts, creature scripts).
-  [ ] Database (SAI, creatures, etc).

This is a proposal following an attempt here (#19671), where bash was showing its limitations when we trying to exclude files from tests with "grep":
* Python is simpler and more familiar to most people
* It has a large number of resources and aids
* It allows more advanced usage for codestyle and various tests

New features: 
* Allow you to exclude one or more files from a test 
* A test does not stop at the first error encountered, and completes all tests to provide a complete list of issues.

## Issues Addressed:

- Closes 

## SOURCE:

The changes have been validated through:
- [ ] Live research (checked on live servers, e.g Classic WotLK, Retail, etc.)
- [ ] Sniffs (remember to share them with the open source community!)
- [ ] Video evidence, knowledge databases or other public sources (e.g forums, Wowhead, etc.)
- [ ] The changes promoted by this pull request come partially or entirely from another project (cherry-pick). **Cherry-picks must be committed using the proper --author tag in order to be accepted, thus crediting the original authors, unless otherwise unable to be found**

## Tests Performed:

This PR has been:
- [X] Tested in-game by the author.
- [ ] Tested in-game by other community members/someone else other than the author/has been live on production servers.
- [ ] This pull request requires further testing and may have edge cases to be tested.

## How to Test the Changes:

- [ ] This pull request can be tested by following the reproduction steps provided in the linked issue
- [ ] This pull request requires further testing. Provide steps to test your changes. If it requires any specific setup e.g multiple players please specify it as well.

1. Add deliberate mistakes (whitespace, blank lines, and others things check by the script)
2. run  python ./apps/codestyle/codestyle.py in the project directory

## Known Issues and TODO List:

- [X] Move in app folder and integrate it github workflow
- [X] Fix whitespace check at the end line of the line
- [X] Fix the situation where we have two blank lines at the end of a file. The final line is not recognised as a blank line.
- [X] Restore the possibility to pass all checks before to return an error
- [X] Improve outputs

## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here on the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

**REMEMBER**: when testing a PR that changes something **generic** (i.e. a part of code that handles more than one specific thing), the tester should not only check that the PR does its job (e.g. fixing spell XXX) but **especially** check that the PR does not cause any regression (i.e. introducing new bugs).

**For example**: if a PR fixes spell X by changing a part of code that handles spells X, Y, and Z, we should not only test X, but **we should test Y and Z as well**.

### CODE REVIEW

**Comment by Grimdhex** [comment]
PR finish. You check the github action to see that the script run fine in the github workflow. For advanced tests just use the command: python ./apps/codestyle/codestyle.py in the source directory

If the change is accepted, I remove the test file. Need also to be merge after https://github.com/azerothcore/azerothcore-wotlk/pull/19735

@Kitzunu

**Comment by sudlud** [comment]
Can you please merge master

**Comment by Kitzunu** [comment]
> Multiple blank line check : Passed
Trailing whitespace check : Passed
GetCounter check : Failed
GetTypeId check : Passed
NpcFlagHelpers check : Failed
Multiple Blank Line check : Failed
Trailing Whitespace check : Failed

The downside to this is that it doesn't tell you where the issues are present

**Comment by Tereneckla** [comment]
> Multiple blank lines found in /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/CMakeLists.txt at line 40
Multiple blank lines found at the end of: /home/runner/work/azerothcore-wotlk/azerothcore-wotlk/src/CMakeLists.txt

Blank lines are specified

**Comment by Grimdhex** [comment]
> > Multiple blank line check : Passed
> > Trailing whitespace check : Passed
> > GetCounter check : Failed
> > GetTypeId check : Passed
> > NpcFlagHelpers check : Failed
> > Multiple Blank Line check : Failed
> > Trailing Whitespace check : Failed
> 
> The downside to this is that it doesn't tell you where the issues are present

As Tereneckla says, it's specified above. But there is a problem with the logic of results output. As all the checks use the same variable (error_handler) as soon as the first check fails, all the others will be indicated as "Failed". I will look this.

File: apps/codestyle/codestyle.py [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+error_handler = False
+results = {
+    "Multiple blank lines check": "Passed",
+    "Trailing whitespace check": "Passed",
+    "GetCounter() check": "Passed",
+    "NpcFlagHelpers check": "Passed"
+}
+
+# Main function to parse all the files of the project
+def parsing_file(directory):
```
Comment (Helias) [inline apps/codestyle/codestyle.py:17]:
```suggestion
def parsing_file(directory: str) -> None:
```

File: apps/codestyle/codestyle.py [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+    for check, result in results.items():
+        print(f"{check} : {result}")
+    if error_handler:
+        print("\nPlease fix the codestyle issues above.")
+        sys.exit(1)
+    else:
+        print(f"\nEverything looks good")
+
+# Codestyle patterns checking for multiple blank lines
+def multipleBlankLines_check(file, file_path):
```
Comment (Helias) [inline apps/codestyle/codestyle.py:44]: I would respect the python snake_case naming here and moreover, let's add the type hints everywhere in any function header please
```suggestion
def multiple_blank_lines_check(file: str, file_path: str) -> None:
```

File: apps/codestyle/codestyle.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    if consecutive_blank_lines >= 1:
+        print(f"Multiple blank lines found at the end of: {file_path}")
+        check_failed = True
+    # Handle the script error and update the result output
+    if check_failed:
+        error_handler = True
+        results["Multiple blank lines check"] = "Failed"
+
+# Codestyle patterns checking for whitespace at the end of the lines
+def trailing_whitespace_check(file, file_path):
```
Comment (Helias) [inline apps/codestyle/codestyle.py:68]:
```suggestion
def trailing_whitespace_check(file: str, file_path: str) -> None:
```

File: apps/codestyle/codestyle.py [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+    for line_number, line in enumerate(file, start = 1):
+        if 'ObjectGuid::GetCounter()' in line:
+            print(f"Please use ObjectGuid::ToString().c_str() instead ObjectGuid::GetCounter(): {file_path} at line {line_number}")
+            if not error_handler:
+                error_handler = True
+                results["GetCounter() check"] = "Failed"
+
+# Codestyle patterns checking for GetTypeId()
+# Unused for now. Need to fix all usage in the core before adding the check to the github workflow
+def getTypeId_check(file, file_path):
```
Comment (Helias) [inline apps/codestyle/codestyle.py:93]: same here, fix the name style and add the type hints please

**Comment by Helias** [comment]
good job by the way 👍

File: apps/codestyle/codestyle.py (hunk 2 above)
Comment (Grimdhex) [reply apps/codestyle/codestyle.py]: file is not really a string but a full text. Are you sure about this type hints ? Pycharm return me a warning and suggest to define file as: TextIO

Comment (Helias) [reply apps/codestyle/codestyle.py]: I guessed it, so may it's wrong, I would trust more Pycharm then hehe, the type depends on what os.walk() returns, and pycharm could get it from there, so should be correct TextIO

**Comment by Helias** [comment]
be sure to fix the codestyle error reports before merging this PR otherwise all the future PRs will fail as well

File: apps/codestyle/codestyle.py [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+    "Trailing whitespace check": "Passed",
+    "GetCounter() check": "Passed",
+    "NpcFlagHelpers check": "Passed"
+}
+
+# Main function to parse all the files of the project
+def parsing_file(directory):
+    for root, _, files in os.walk(directory):
+        for file in files:
+            if not file.endswith('.ico'):                                   # Skip .ico files that cannot be read
```
[the remark this answers is not in the document]
Comment (Grimdhex) [reply apps/codestyle/codestyle.py:22]: With my attempts, I'm only saving a line or two and I'm experiencing access error problems. I suggest looking at optimisation for a future PR

**Comment by Grimdhex** [comment]
If it's fine for you, the PR is ready to be merged.

**Review by Helias — APPROVED** [review]

**Pull request merged**